Describe new syntax for hosted dependencies (#3647)
dart-lang/site-www: src/tools/pub/dependencies.md

{% endprettify %}

This YAML code creates a dependency on the `transmogrify` package
using the default package repository ([pub.dev]({{site.pub}})) and
allowing any version from `1.0.0` to `2.0.0` (but not including `2.0.0`).
See the [version constraints](#version-constraints)
section of this page for syntax details.

If you want to specify a custom package repository,

> use the following syntax:

Reviewer: I'm uneasy about showing the 2.15 syntax here before 2.15 has a stable channel release. One possible solution is to remove the example and do a forward link to #hosted-packages.

Reviewer: Do we know which syntax people are most likely to use? Which they're most likely to have trouble with?

Reviewer: Finally... Will the pre-2.15 syntax still work in 2.15+?

Author: > I don't want to show the 2.15 syntax here before 2.15 has a stable channel release.

Good point. I guess the first example is mainly to explain the concept of a description. We could instead let that be a git or path dependency... Or just waiting merging this until the 2.15 release.

> Do we know which syntax people are most likely to use? Which they're most likely to have trouble with?

Hopefully the new syntax! We're adding much better support for third-party repositories in 2.15, so it gets much more interesting to specify.

> Finally... Will the pre-2.15 syntax still work in 2.15+?

Yes it will.

{% prettify yaml tag=pre+code %}
dependencies:
  transmogrify:
    hosted: 'http://some-package-server.com'

>     version: ^1.0.0
> {% endprettify %}

Reviewer 2: ```suggestion
    version: ^1.0.0
environment:
  sdk: >= 2.15.0 <3.0.0
{% endprettify %}
```

* [Git packages](#git-packages)
* [Path packages](#path-packages)

### Hosted packages

A *hosted* package is one that can be downloaded from the pub.dev site
(or another HTTP server that speaks the same API). Here's an example
of declaring a dependency on a hosted package:

```yaml
dependencies:
  transmogrify: ^1.4.0
```

This example specifies that your package depends on a hosted package named
`transmogrify` and works with any version from 1.4.0 to 2.0.0
(but not 2.0.0 itself).

If you want to use your own package repository,
you can use `hosted` to specify its URL.
The following YAML code creates a dependency on the `transmogrify` package
using the `hosted` source.

{% prettify yaml tag=pre+code %}
environment: 
  sdk: >=[!2.15.0!] < 3.0.0

dependencies:
  transmogrify:

>     [!hosted: http://some-package-server.com!]

Reviewer 3: With the new `https://dart.dev/tools/linter-rules#secure_pubspec_urls` linter rule, perhaps we should use `https` links in the examples.

```suggestion
    [!hosted: https://some-package-server.com!]
```

    version: ^1.4.0
{% endprettify %}

The version constraint is optional but recommended. If no version constraint is
given, `any` is assumed.

{{site.alert.version-note}}
If your package has a [language version][] before 2.15,
you must use a more verbose `hosted` format:

{% prettify yaml tag=pre+code %}
environment:
  sdk: >=[!2.14.0!] < 3.0.0

dependencies:
  transmogrify:
    [!hosted:!]
      [!name: transmogrify!]

>       [!url: http://some-package-server.com!]

Reviewer 3: Same here.

```suggestion
      [!url: https://some-package-server.com!]
```

The syntax looks like this:

> 

Author: Done

```yaml
dependencies:
  flutter_driver:
    sdk: flutter
```

The identifier after `sdk:` indicates which SDK the package comes from.
If it's `flutter`, the dependency is satisfiable as long as:

* Pub is running in the context of the `flutter` executable
* The Flutter SDK contains a package with the given name

If it's an unknown identifier, the dependency is always considered unsatisfied.

> A package with an `sdk` dependency

Author: When do we remove documentation for old versions? Version 1.19 starts to be pretty niche...

`transmogrify` and will work with any version from 1.4.0 to 2.0.0
(but not 2.0.0 itself).

If you want to use your own package repository, you can use a description that
specifies its URL:

{% prettify yaml tag=pre+code %}
# This syntax requires language version 2.15 or higher
environment: 
  sdk: >=2.15.0 < 3.0.0

dependencies:
  transmogrify:
    hosted: http://some-package-server.com
    version: ^1.4.0
{% endprettify %}

> Before Dart language version 2.15, the format is different:

Reviewer 3: Depending on when this PR is merged, do we want to separate the old version with a version-note mentioning language version like we do in https://github.com/dart-lang/site-www/blob/master/src/_guides/language/analysis-options.md?plain=1#L195-L201? To be consistent with elsewhere on the website. Then in the currently supported version above, don't add the "# This syntax requires language version 2.15 or higher" comment?

Author: Done

[connect to the repo using SSH,][GitHub SSH]
then you can depend on the package by using the repo's SSH URL:

```yaml
dependencies:
  kittens:
    git: git@github.com:munificent/kittens.git
```

If you want to depend on a specific commit, branch, or tag,
add a `ref` key to the description:

```yaml
dependencies:
  kittens:
    git:
      url: git@github.com:munificent/kittens.git
      ref: some-branch
```

The ref can be anything that Git allows to [identify a commit.][commit]

[commit]: https://www.kernel.org/pub/software/scm/git/docs/user-manual.html#naming-commits

Pub assumes that the package is in the root of the Git repository. To specify a
different location in the repo, specify a `path` relative to the repository
root:

```yaml
dependencies:
  kittens:
    git:
      url: git@github.com:munificent/cats.git
      path: path/to/kittens
```

The path is relative to the Git repo's root.

> Git dependencies are not allowed on [pub.dev][pubsite].

Reviewer 3: I wonder if it would be more clear if we specify in what way they are not allowed. What do you think?

```suggestion
Git dependencies are not allowed as dependencies
for packages uploaded to [pub.dev][pubsite].
```

Author: Yeah - that's better!

Pub assumes that the package is in the root of the Git repository.

> To specify a different location in the repo, use the `path` argument:

Author: Done
